### PR TITLE
container: init service for linux and mac

### PR DESCRIPTION
**Pull request by michaelCTS**
### Description

Evolved from the original PR based on https://github.com/nix-community/home-manager/pull/2630 by @MaeIsBad, this PR introduces 3 major options:

**`virtualisation.containers`**

From the original PR and nixpkgs, allows managing `regstristries.conf` (used by podman and others, but not docker) declaratively. `containers.conf` is modified by the aforementioned tools and managing by `home-manager` is considered out of scope of this PR.

**`virtualisation.oci-containers`**

Inspired by [`nixpkgs`](https://github.com/NixOS/nixpkgs/blob/db9208ab987cdeeedf78ad9b4cf3c55f5ebd269b/nixos/modules/virtualisation/oci-containers.nix#L335-L341), a simple method to enable containers which will pick the appropriate/supported oci-container implementation of your platform. ATM only linux and mac are supported by activating `virtualisation.podman`.

`virtualisation.oci-containers.enable = true` and you're off.

**`virtualisation.podman`**

Enables support for podman on linux and mac. Linux is more simple as it activates as systemd user service and that's that.

Mac is the difficult child in this relationship as it requires a VM to function. I opted to support basic, declarative VM management using `virtualisation.podman.machines.<name>`. `podman` stores its VM configs in JSON and outputs them using the CLI in JSON as well. Since python is my forté and not `bash` + `jq`, I wrote a helper called `podmactl` which is called at activation of a new home-manager generation.

It diffs the existing machines from requested machines (JSON generated by `nix`) and the required machines are generated and only one is active (podman doesn't allow multiple running machines).

### Checklist

- [x] Change is backwards compatible.
- [x] Code formatted with `./format`.
- [x] Code tested through `nix-shell --pure tests -A run.all` or `nix develop --ignore-environment .#all` using Flakes.
- [x] Test cases updated/added. See [example](https://github.com/nix-community/home-manager/commit/f3fbb50b68df20da47f9b0def5607857fcc0d021#diff-b61a6d542f9036550ba9c401c80f00ef).
- [ ] Commit messages are formatted like

    ```
    {component}: {description}

    {long description}
    ```

    See [CONTRIBUTING](https://github.com/nix-community/home-manager/blob/master/docs/contributing.adoc#sec-commit-style) for more information and [recent commit messages](https://github.com/nix-community/home-manager/commits/master) for examples.

- If this PR adds a new module

  - [x] Added myself as module maintainer. See [example](https://github.com/nix-community/home-manager/blob/068ff76a10e95820f886ac46957edcff4e44621d/modules/programs/lesspipe.nix#L6).

### CODE REVIEW

**Comment by MaeIsBad** [comment]
I'd be glad to review this once it gets out of the draft stage and help with maintenance.

One thing to note is that for a while the regular nixos `virtualisation.podman.enable` option already installs a user systemd socket you can use

File: modules/lib/maintainers.nix [hunk 1]
```diff
@@ -211,6 +211,12 @@
     githubId = 1472826;
     name = "Max Smolin";
   };
+  michaelCTS = {
+    email = "michael.vogel@cts.co";
+    github = "michaelCTS";
+    githubId = 132582212;
+    name = "Michael Vogel";
+  };
```
Comment (rycee) [inline modules/lib/maintainers.nix:219]: Since you are already listed in Nixpkgs' maintainer-list.nix you don't need to be listed here as well.

File: modules/virtualisation/containers.nix [hunk 2]
```diff
@@ -0,0 +1,89 @@
+{ config, lib, pkgs, ... }:
+let
+  cfg = config.virtualisation.containers;
+
+  inherit (lib) mkOption types;
+
+  toml = pkgs.formats.toml { };
+in {
+  meta.maintainers = [ lib.hm.maintainers.bad ];
```
Comment (rycee) [inline modules/virtualisation/containers.nix:9]: There is no such maintainer listed anywhere, as far as I can see?

File: modules/virtualisation/containers.nix [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+  options.virtualisation.containers = {
+
+    enable = lib.mkEnableOption "the common containers configuration module";
+
+    ociSeccompBpfHook.enable = lib.mkEnableOption "the OCI seccomp BPF hook";
+
+    containersConf.settings = mkOption {
+      type = toml.type;
+      default = { };
+      description = "containers.conf configuration";
```
Comment (rycee) [inline modules/virtualisation/containers.nix:19]:
```suggestion
      description = "The {file}`containers.conf` configuration.";
```

File: modules/virtualisation/containers.nix [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+            docker-daemon = {
+              "" = [ { type = "insecureAcceptAnything"; } ];
+            };
+          };
+        }
+      '';
+      description = ''
+        Signature verification policy file.
+        If this option is empty the default policy file from
+        <literal>skopeo</literal> will be used.
```
Comment (rycee) [inline modules/virtualisation/containers.nix:64]: We now use [Nixpkgs' Markdown](https://nixos.org/manual/nixpkgs/unstable/#sec-contributing-markup) for descriptions so need someting like:

```suggestion
        `skopeo` will be used.
```

File: modules/virtualisation/podman/podman.nix [hunk 5]
```diff
@@ -0,0 +1,165 @@
...
+  options.services.podman = {
+
+    enable = mkOption {
+      type = types.bool;
+      default = false;
+      description = ''
+        This option enables Podman, a daemonless container engine for
+        developing, managing, and running OCI Containers on your Linux System.
+
+        It is a drop-in replacement for the <command>docker</command> command.
```
Comment (rycee) [inline modules/virtualisation/podman/podman.nix:45]: Isn't this true only if `dockerCompat = true`?

File: modules/virtualisation/podman/podman.nix [hunk 6]
```diff
@@ -0,0 +1,165 @@
...
+
+in {
+  meta.maintainers = [ lib.maintainers.michaelCTS ];
+  imports = [
+    (lib.mkRenamedOptionModule [ "virtualisation" "podman" "libpod" ] [
+      "virtualisation"
+      "containers"
+      "containersConf"
+    ])
+  ];
```
Comment (rycee) [inline modules/virtualisation/podman/podman.nix:35]: There has never been such an option before so no need to do anything here:
```suggestion

```

File: modules/virtualisation/podman/podman.nix [hunk 7]
```diff
@@ -0,0 +1,165 @@
...
+    } ''
+      mkdir -p $out/bin
+      ln -s ${podmanPackage}/bin/podman $out/bin/docker
+
+      mkdir -p $man/share/man/man1
+      for f in ${podmanPackage.man}/share/man/man1/*; do
+        basename=$(basename $f | sed s/podman/docker/g)
+        ln -s $f $man/share/man/man1/$basename
+      done
+    '';
```
Comment (rycee) [inline modules/virtualisation/podman/podman.nix:25]: This seems like it should be a feature of the podman package in Nixpkgs, similar to how the yt-dlp package has a `withAlias` argument. For example,

``` nix
pkgs.podman.override { inherit (cfg) extraPackages enableDockerAlias; }
```

Perhaps it would be possible to add this functionality to the podman package in Nixpkgs?

File: modules/virtualisation/podman/podman.nix [hunk 8]
```diff
@@ -0,0 +1,165 @@
...
+      default = false;
+      description = ''
+        Make the Podman socket available in place of the Docker socket, so
+        Docker tools can find the Podman socket.
+
+        Podman implements the Docker API.
+      '';
+    };
+
+    dockerCompat = mkOption {
```
Comment (rycee) [inline modules/virtualisation/podman/podman.nix:60]: Perhaps can have something more descriptive? E.g.,

```suggestion
    enableDockerAlias = mkOption {
```

File: modules/virtualisation/podman/podman.nix [hunk 9]
```diff
@@ -0,0 +1,165 @@
...
+      default = false;
+      description = ''
+        This option enables Podman, a daemonless container engine for
+        developing, managing, and running OCI Containers on your Linux System.
+
+        It is a drop-in replacement for the <command>docker</command> command.
+      '';
+    };
+
+    dockerSocket = mkOption {
```
Comment (rycee) [inline modules/virtualisation/podman/podman.nix:49]: I think prefixing with "enable" will make it a bit clearer:

```suggestion
    enableDockerSocket = mkOption {
```

File: modules/virtualisation/podman/podman.nix [hunk 10]
```diff
@@ -0,0 +1,165 @@
...
+        [
+          pkgs.gvisor
+        ]
+      '';
+      description = ''
+        Extra packages to be installed in the Podman wrapper.
+      '';
+    };
+
+    package = lib.mkOption {
```
Comment (rycee) [inline modules/virtualisation/podman/podman.nix:89]: Typically we call this type of option `finalPackage` (and add `readOnly = true`):
```suggestion
    finalPackage = lib.mkOption {
```

**Comment by rycee** [comment]
Thanks a lot! It would be great to have this functionality. I started adding a few comments. Will continue when I find some more time.

File: modules/virtualisation/podman/podman.nix (hunk 5 above)
Comment (MaeIsBad) [reply modules/virtualisation/podman/podman.nix:45]: Podman is a drop in, docker compat just creates an alias binary for script compatibility.

Comment (rycee) [reply modules/virtualisation/podman/podman.nix:45]: Ok, I think it may be my misinterpretation of "drop-in". I was thinking that it would be a drop-in in the sense that if you today have a script that uses the docker command, then after switching you could use the script unchanged with Podman. But it actually is drop-in in the sense that you can use the script with the only change being s/docker/podman/ 🙂

So no need to change, except for the conversion to Markdown
```suggestion
        It is a drop-in replacement for the {command}`docker` command.
```

File: modules/virtualisation/containers.nix (hunk 2 above)
Comment (MaeIsBad) [reply modules/virtualisation/containers.nix:9]: I think this is a leftover from my original PR, where I was the only maintainer. 
I don't feel confident in being the sole maintainer but I would be happy to become a second maintainer for this module.

**Comment by michaelCTS** [comment]
Thanks for the review @rycee . For now, I made only minimal changes after recovering @MaeIsBad's PR. There might be bigger changes coming tomorrow (I get to work on this once a week) as I have a look at how `virtualisation.podman.enable` currently works on nixpkgs. CNI isn't used for networking anymore and there may be more changes to incorporate.

File: modules/lib/maintainers.nix (hunk 1 above)
Comment (michaelCTS) [reply modules/lib/maintainers.nix]: Replace all references with `pkgs.lib.maintainers.michaelCTS`

File: modules/virtualisation/containers.nix (hunk 2 above)
Comment (michaelCTS) [reply modules/virtualisation/containers.nix]: I'll let you decide on that once the PR is ready @MaeIsBad . Then you can have an idea of what you're getting yourself into :grin:

**Comment by MaeIsBad** [comment]
With flakes you can point the flake input by using `git+file:///home/user/home-manager`. That's how I test changes to home-manager

File: modules/virtualisation/podman/podman.nix (hunk 7 above)
Comment (michaelCTS) [reply modules/virtualisation/podman/podman.nix]: It's a good idea. I'll let somebody else handle that due to a lack of time on my end :innocent:

**Comment by michaelCTS** [comment]
Thanks @MaeIsBad . I don't use flakes, so had to find another way. Turned out it was

```sh
cd $proj
nix-shell -A install
home-manager -I home-manager=$proj switch
```

The PR is ready for review. I'll leave comments where I'm unsure about stuff.

As for the commit messages: only noticed those while checking of the list. What to do? Just squash everything? I tried to add many comments in the code, so a huge commit might be acceptable?
Otherwise I can rebase and reword each commit. Whatever is prefered.

File: modules/virtualisation/containers.nix [hunk 11]
```diff
@@ -0,0 +1,87 @@
...
+
+    ociSeccompBpfHook.enable = lib.mkEnableOption "the OCI seccomp BPF hook";
+
+    # Unimplemented as podman modifies the file and it would have to be merged
+    #  not replaced when activating a new config
+    #    containersConf.settings = mkOption {
+    #      type = toml.type;
+    #      default = { };
+    #      description = "The {file}`containers.conf` configuration.";
+    #    };
```
Comment (michaelCTS) [inline modules/virtualisation/containers.nix:22]: Present in `nixpkgs` https://github.com/NixOS/nixpkgs/blob/230262cda5dc5b704aae91273e2676a9ac65c81e/nixos/modules/virtualisation/containers.nix#L31-L35


Podman modifies the files in the user's home, outside of `home-manager`. Fixing that in this PR is out of scope, so not sure if this comment (and related commented out code) should stay or just be removed.

File: modules/virtualisation/podman/podmactl/podmactl.py
Comment (michaelCTS) [inline modules/virtualisation/podman/podmactl/podmactl.py:1]: Not sure where to put this. Wanted unit testing so putting it in the `podman.nix` didn't seem right.

File: modules/virtualisation/podman/podman.nix [hunk 12]
```diff
@@ -0,0 +1,276 @@
...
+      '';
+    };
+
+    #    enableNvidia = mkOption {
+    #      type = types.bool;
+    #      default = false;
+    #      description = ''
+    #        Enable use of NVidia GPUs from within podman containers.
+    #      '';
+    #    };
```
Comment (michaelCTS) [inline modules/virtualisation/podman/podman.nix:116]: Requires `virtualisation.containers.containersConf.settings` which is a hassle to implement and out of scope (see comment on the options in `containers.nix`). Remove comment or keep?

File: modules/virtualisation/podman/podman.nix [hunk 13]
```diff
@@ -0,0 +1,276 @@
...
+        enable = true; # Enable common /etc/containers configuration
+        #        containersConf.settings = lib.optionalAttrs cfg.enableNvidia {
+        #          engine = {
+        #            conmon_env_vars =
+        #              [ "PATH=${lib.makeBinPath [ pkgs.nvidia-podman ]}" ];
+        #            runtimes.nvidia =
+        #              [ "${pkgs.nvidia-podman}/bin/nvidia-container-runtime" ];
+        #          };
+        #        };
+      };
```
Comment (michaelCTS) [inline modules/virtualisation/podman/podman.nix:194]: Requires `virtualisation.containers.containersConf.settings` which is a hassle to implement and out of scope (see comment on the options in `containers.nix`).
Delete comment or keep?

File: modules/virtualisation/podman/podman.nix [hunk 14]
```diff
@@ -0,0 +1,276 @@
...
+          image_path = "fedora-coreos-38.20230806.3.0-qemu.x86_64.qcow2.xz";
+          qemu_binary = "$${pkgs.qemu}/bin/qemu-system-x86_64";
+        };
+      };
+      # One and only one machine may be active at any given time
+      apply = machines:
+        assert ((lib.lists.count (machine: machine.active)
+          (lib.attrsets.attrValues machines)) == 1);
+        machines;
+    };
```
Comment (michaelCTS) [inline modules/virtualisation/podman/podman.nix:175]: Unsure how to do this with a proper error message e.g `Only one machine may be active at a time`

File: modules/virtualisation/podman/podmactl/.gitignore [hunk 15]
```diff
@@ -0,0 +1 @@
+result
```
Comment (MaeIsBad) [inline modules/virtualisation/podman/podmactl/.gitignore:1]: No other modules need to add a .gitignore file, I don't see why this one should. You could maybe put this it into the top level gitignore file

File: modules/virtualisation/podman/podman.nix [hunk 16]
```diff
@@ -0,0 +1,276 @@
...
+    # options here are loaded into python
+    # for simplicity, please use snake_case
+    options = {
+      active = mkOption {
+        type = types.bool;
+        default = false;
+        description =
+          "This machine should be started. Only one machine can be active at a time";
+      };
+      qemu_binary = mkOption {
```
Comment (MaeIsBad) [inline modules/virtualisation/podman/podman.nix:38]: nix options are typically written using camelCase, same for the rest of this file

File: modules/virtualisation/podman/podman.nix [hunk 17]
```diff
@@ -0,0 +1,276 @@
...
+              After = "podman.socket";
+              Documentation = "man:podman-system-service(1)";
+              StartLimitIntervalSec = 0;
+            };
+            Service = {
+              Type = "exec";
+              KillMode = "process";
+              Environment = ''LOGGING=" --log-level=info"'';
+              ExecStart =
+                [ "" "${cfg.finalPackage}/bin/podman $LOGGING system service" ];
```
Comment (MaeIsBad) [inline modules/virtualisation/podman/podman.nix:214]:
```suggestion
                [ "${cfg.finalPackage}/bin/podman" "$LOGGING" "system" "service" ];
```

See my original PR(https://github.com/nix-community/home-manager/pull/2630#discussion_r797006402)

File: modules/virtualisation/podman/podman.nix [hunk 18]
```diff
@@ -0,0 +1,276 @@
...
+            Install.WantedBy = [ "sockets.target" ];
+          };
+
+        };
+      }
+      (mkIf cfg.enableDockerSocket {
+        home.sessionVariables."DOCKER_HOST" =
+          "unix:///run/user/$UID/podman/podman.sock";
+      })
+    ]))
```
Comment (MaeIsBad) [inline modules/virtualisation/podman/podman.nix:238]:
```suggestion
      (mkIf cfg.enableDockerSocket {
        home.sessionVariables."DOCKER_HOST" =
          "unix:///$XDG_RUNTIME_DIR/podman/podman.sock";
      })
    ]))
```

https://github.com/nix-community/home-manager/pull/2630#discussion_r810568878

File: modules/virtualisation/podman/podmactl/README.md
Comment (MaeIsBad) [inline modules/virtualisation/podman/podmactl/README.md:1]: This feels out of scope for a home-manager module, maybe you should consider moving this to a separate project..? Up to the home-manager devs to decide

File: modules/virtualisation/containers.nix (hunk 11 above)
Comment (michaelCTS) [reply modules/virtualisation/containers.nix:22]: I'll give that a try and see if podman complains :)

File: modules/virtualisation/oci-containers.nix [hunk 19]
```diff
@@ -0,0 +1,31 @@
...
+  inherit (lib) mkDefault mkIf mkMerge mkOption types;
+
+  defaultBackend = "podman";
+in {
+
+  meta.maintainers = [ pkgs.lib.maintainers.michaelCTS ];
+  options.virtualisation.oci-containers = {
+
+    enable = lib.mkEnableOption
+      "a convenience option to enable containers in platform-agnostic manner";
```
[the remark this answers is not in the document]
Comment (michaelCTS) [reply modules/virtualisation/oci-containers.nix:16]: This is taken from `nixpkgs` https://github.com/NixOS/nixpkgs/blob/ec68439bdc77ee684a4b2846f03677c5c50920a9/nixos/modules/virtualisation/oci-containers.nix#L335-L349

The problem with `containers` is that it used by `podman`. I'm not versed enough in nix and nixpkgs to not create a recursion problem with `containers.enable` using `containers.backend` to enable `podman` which itself then sets `containers.enable` and configures the settings and other stuff. I gave it a shot and immediately ran into recursion errors.

Since `nixpkgs` doesn't bother with that, I erred on the side of "they probably know better what they're doing that some 1 year nix scrub like me".

File: modules/virtualisation/podman/podmactl/.gitignore (hunk 15 above)
Comment (michaelCTS) [reply modules/virtualisation/podman/podmactl/.gitignore:1]: Reasonable suggestion.

File: modules/virtualisation/podman/podman.nix (hunk 16 above)
Comment (michaelCTS) [reply modules/virtualisation/podman/podman.nix:38]: This is specifically because it's exported into a JSON which is used by the python script to apply the machine config. There's a comment a few lines up that mentions this.

I'll have a look at the rest of the file and see if I didn't it respect it outside of machine options.

**Comment by rycee** [comment]
Finally had a chance to look at this and overall looks very nice. I made some minor cleanups and squashed the commits but unfortunately was not allowed to push to this branch. Would you mind resetting the PR to https://github.com/nix-community/home-manager/commit/faa4b16358ebd8607499538dbea02bfba5930bd9

The only thing I'm concerned about is the inclusion of the Python project, I would feel more comfortable with it being a separate project and pulled from Nixpkgs. Then I imagine it could be used elsewhere as well, perhaps in nix-darwin or devenv?

File: modules/virtualisation/podman/podmactl/README.md
Comment (rycee) [reply modules/virtualisation/podman/podmactl/README.md:1]: I agree it should be moved into its own repository, if you don't want it under your own account, then perhaps it could be hosted under nix-community, since it may be useful for other Nix related projects?

**Comment by michaelCTS** [comment]
Thanks for the review @rycee and happy new year!

If it's not an issue, I'll move the python code out to a repo I own and later try to get it into nixpkgs or nix-community or something. My experience with nixpkgs hasn't been good as PRs seem to take a considerable amount of time to be merged if they are merged at all, and I don't want that to hold up this PR.

My mac isn't with me at the moment, so the earliest I can test the changes will be this Sunday, 2024-01-07.

Cheers

File: modules/misc/news.nix [hunk 20]
```diff
@@ -1348,14 +1348,24 @@ in
           A new module is available: 'programs.gradle'.
         '';
       }
-
+
```
Comment (terlar) [inline modules/misc/news.nix:1351]: Noticed some dead white-space here

File: modules/virtualisation/podman.nix [hunk 21]
```diff
@@ -0,0 +1,279 @@
...
+            Service = {
+              Type = "exec";
+              KillMode = "process";
+              Environment = ''LOGGING=" --log-level=info"'';
+              ExecStart = [
+                "${cfg.finalPackage}/bin/podman"
+                "$LOGGING"
+                "system"
+                "service"
+              ];
```
Comment (rycee) [inline modules/virtualisation/podman.nix:215]: This will cause the service file to contain one ExecStart for each entry in the list. I assume you want to run this as a single command?

```suggestion
              ExecStart = lib.concatStringsSep " " [
                "${cfg.finalPackage}/bin/podman"
                "$LOGGING"
                "system"
                "service"
              ];
```

File: modules/virtualisation/podman.nix [hunk 22]
```diff
@@ -0,0 +1,279 @@
...
+              Requires = "podman.socket";
+              After = "podman.socket";
+              Documentation = "man:podman-system-service(1)";
+              StartLimitIntervalSec = 0;
+            };
+
+            Service = {
+              Type = "exec";
+              KillMode = "process";
+              Environment = ''LOGGING=" --log-level=info"'';
```
Comment (rycee) [inline modules/virtualisation/podman.nix:209]: Generally speaking it is best to make the `Environment` attributes have list values to allow for merging:

```suggestion
              Environment = [ ''LOGGING="--log-level=info"'' ];
```

**Comment by rycee** [comment]
Sorry about the delay. I had a chance to try this out. But I don't know how to get it to work. Firstly, the generated systemd service file seems to be invalid due to generating multiple ExecStart fields. Once I turned into a single command I get an error

    Error: command required for rootless mode with multiple IDs:
           exec: "newuidmap": executable file not found in $PATH

I didn't do any special configuration, just added

    virtualisation.podman.enable = true;

in my HM configuration. Do I need some additional configuration to get the service up?

**Comment by michaelCTS** [comment]
I'm sorry, I'll have to close this. The [podman update to 5.0](https://blog.podman.io/2024/03/podman-5-0-has-been-released/) made changes to `podman machine` that I'd have to integrate and I've since moved away from trying to support Mac as my company has another solution.

Anybody is free to try and take this over. I'll leave my branch up.

**Comment by jlbribeiro** [comment]
@michaelCTS Thank you so so much for your work on this PR (and the reviewers for their feedback on it)!

My humble suggestion/request for whoever picks up this wonderful work (unless #4801 gets merged before): I think macOS/Darwin support should _not_ block the merge of this module, as the [original @MaeIsBad PR this one is based off](https://github.com/nix-community/home-manager/pull/2630) and #4336 indicate, this has been a need for a while (at least for me).

Imho, NixOS should be the priority, given it would enable proper rootless `podman`, since https://github.com/NixOS/nixpkgs/issues/207050 and https://github.com/NixOS/nixpkgs/issues/259770 are open, and the discussion in https://github.com/containers/podman/discussions/20573 seems to point in the direction of using `systemd --user` units, and not system `systemd` + `User=` (previously discussed in https://github.com/containers/podman/issues/12778).

After supporting NixOS, generic Linux systems, and only then, macOS/Darwin (which will always be a bit more complex). Linux-only support would already be a great improvement on QoL regarding `podman` on Nix-managed systems.
(With that said, taking the time to design an API that is extension-friendly (so as to allow supporting other platforms) is always good.)
